### PR TITLE
Convert client into a REST API

### PR DESCRIPTION
**Pull request by dorranh**
This PR includes a variety of changes for converting the monolithic Flask `client` into a REST API which is only concerned with serving data. #37 introduces a new front-end which will utilize this updated API. While making the changes, I've also included an additional `packages` endpoint which can be used to list the packages available in the graph.

As a part of these changes, I've introduced a Nix flake which provides some of the project's dev dependencies. As discussed internally, I have also ported the Flask API to FastAPI to allow for better type checking and application life cycle management. Finally, I made some small updates to the documentation to reflect the changes to the project.

Note that I have introduced a few standard dev tools for the Python code, namely black, isort, flake8, and pyright. These can be run via the provided Nix flake commands in the `api` directory, and we can enforce them in CI as well (perhaps in a follow-up PR to avoid this one growing any larger).

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,103 @@
+{
+  description = "nixpkgs-graph-explorer flake";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+    nixpkgs_old_cypress.url =
+      "github:NixOS/nixpkgs?rev=bf972dc380f36a3bf83db052380e55f0eaa7dcb6";
+    flake-utils.url = "github:numtide/flake-utils";
+    devshell = {
+      url = "github:numtide/devshell";
```
Comment (jlesquembre) [inline flake.nix:10]: I think we should use devenv instead. There are plans to deprecate `devshell`: https://github.com/numtide/devshell/issues/242#issuecomment-1464001725
Plus with `devenv` we should not need docker for development. 

I can help with it, probably it's better to create a new PR after this gets merged.

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,103 @@
+{
+  description = "nixpkgs-graph-explorer flake";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+    nixpkgs_old_cypress.url =
```
[the remark this answers is not in the document]
Comment (dorranh) [reply flake.nix:6]: Ah good catch, this was from the front-end changes I had been making on the original branch.

File: flake.nix (hunk 1 above)
Comment (dorranh) [reply flake.nix:8]: https://github.com/tweag/nixpkgs-graph-explorer/issues/40

**Comment by dorranh** [comment]
@jlesquembre, thanks for the review - I've pushed the requested changes. Let me know if you or @GuillaumeDesforges have any other feedback or if this is good to go.

**Comment by GuillaumeDesforges** [comment]
I'd rather merge it asap since it's a big change and #37 depends on it.

**Comment by dorranh** [comment]
Sounds good, thanks @GuillaumeDesforge.

**Pull request merged**